### PR TITLE
[RFC] Fix location list updates.

### PR DESCRIPTION
**Pull request by llorens**
[reproduce.txt](https://github.com/neovim/neovim/files/1347011/reproduce.txt)
Fix quickfix performance optimization which prevented quickfix items
from being updated when there were multiple windows with location lists
but the buffer with errors only in one of the lists.

Attached a script reproducing the issue.

### CODE REVIEW

**Comment by KillTheMule** [comment]
Seems like one of the tests isn't happy. I'll leave the error message, since travis' logs are kinda annoying to search through:

```
From test_quickfix.vim:
Found errors in Test_adjust_lnum():
function RunTheTest[13]..Test_adjust_lnum[2]..Xadjust_qflnum line 21: Expected 13 but got 11
```

Could your script be made into a test? If you need help with that, let me know :)

**Comment by justinmk** [comment]
Thanks, are you sure this isn't covered by an upstream vim patch? Would be better to merge those instead. They will also contain test coverage.

**Comment by llorens** [comment]
Somehow I managed to fix one problem by introducing another one. Yes, I did not run the tests - sorry.
Regarding a fix upstream: there is none so far.

**Comment by llorens** [comment]
@KillTheMule I can work on converting it into a test but I have very little time and I don't even know where to put it (would test/functional/ui/quickfix_spec.lua be a good place?). I assume src/nvim/testdir/test_quickfix.vim is not the right place since these are Vim's tests and Neovim has its own test system.

**Comment by KillTheMule** [comment]
`test/functional/ui/quickfix_spec.lua ` sounds quite right ;)

**Comment by chrisbra** [comment]
> Regarding a fix upstream: there is none so far.

Then perhaps you should raise an issue there?

**Comment by llorens** [comment]
@KillTheMule I figured the test has nothing to do with the UI so I put it in the core directory (quickfix is core functionality, right?). Yes, I am low on time, but this was a good chance to look at the testing framework in Neovim.

File: test/functional/core/quickfix_spec.lua [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	  \ 'text': "This is the error line.",
+	  \ }
+	let qf_item['bufnr'] = bufnr('%')
+	call setloclist(0, [qf_item])
+	wincmd p
+	let qf_item['bufnr'] = bufnr('%')
+	call setloclist(0, [qf_item])
+	1del _
+	call append(0, ['New line 1', 'New line 2', 'New line 3'])
+	silent ll
```
Comment (KillTheMule) [inline test/functional/core/quickfix_spec.lua]: This should be indented for better readability.

**Comment by KillTheMule** [comment]
I don't really mind, but since there already was a file for quickfix testing, that seemed the canonical place to put it. The testing framework is awesome alright :)

File: test/functional/core/quickfix_spec.lua (hunk 1 above)
Comment (llorens) [reply test/functional/core/quickfix_spec.lua]: Yes, these were tabs. Sorry.

**Comment by justinmk** [comment]
@llorens If I revert the changes to mark.c the test doesn't fail.

**Comment by llorens** [comment]
@justinmk You have to revert the changes in quickfix.c as well. In fact I pushed the change as 2 separate commits: 1st one introduces a failing test and the 2nd one fixes the problem and makes the test succeed. Can you verify by simply checking out the 1st of the commits?

**Pull request merged**